examples/net/mqttclient/natiu/main.go

```diff
@@ -0,0 +1,135 @@
+// This example is an MQTT client built with the natiu-mqtt package.  It sends
+// machine.CPUFrequency() readings to the broker every second for 10 seconds.
+//
+// Note: It may be necessary to increase the stack size when using
+// paho.mqtt.golang.  Use the -stack-size=4KB command line option.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"machine"
+	"math/rand"
+	"net"
+	"time"
+
+	mqtt "github.com/soypat/natiu-mqtt"
+)
+
+var (
+	ssid   string
+	pass   string
+	broker string = "test.mosquitto.org:1883"
+	topic  string = "cpu/freq"
+)
+
+func main() {
+	waitSerial()
+
+	if err := netdev.NetConnect(); err != nil {
+		log.Fatal(err)
+	}
+
+	clientId := "tinygo-client-" + randomString(10)
+	fmt.Printf("ClientId: %s\n", clientId)
+
+	// Get a transport for MQTT packets
+	fmt.Printf("Connecting to MQTT broker at %s\n", broker)
+	conn, err := net.Dial("tcp", broker)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer conn.Close()
+
+	// Create new client
+	client := mqtt.NewClient(mqtt.ClientConfig{
+		Decoder: mqtt.DecoderNoAlloc{make([]byte, 1500)},
+		OnPub: func(_ mqtt.Header, _ mqtt.VariablesPublish, r io.Reader) error {
+			message, _ := io.ReadAll(r)
+			fmt.Printf("Message %s received on topic %s\n", string(message), topic)
+			return nil
+		},
+	})
+
+	// Connect client
+	var varconn mqtt.VariablesConnect
+	varconn.SetDefaultMQTT([]byte(clientId))
+	ctx, _ := context.WithTimeout(context.Background(), 10*time.Second)
+	err = client.Connect(ctx, conn, &varconn)
+	if err != nil {
+		log.Fatal("failed to connect: ", err)
+	}
+
+	// Subscribe to topic
+	ctx, _ = context.WithTimeout(context.Background(), 10*time.Second)
+	err = client.Subscribe(ctx, mqtt.VariablesSubscribe{
+		PacketIdentifier: 23,
+		TopicFilters: []mqtt.SubscribeRequest{
+			{TopicFilter: []byte(topic), QoS: mqtt.QoS0},
+		},
+	})
+	if err != nil {
+		log.Fatal("failed to subscribe to", topic, err)
+	}
+	fmt.Printf("Subscribed to topic %s\n", topic)
+
+	// Publish on topic
+	pubFlags, _ := mqtt.NewPublishFlags(mqtt.QoS0, false, false)
+	pubVar := mqtt.VariablesPublish{
+		TopicName: []byte(topic),
+	}
+
+	for i := 0; i < 10; i++ {
+		if !client.IsConnected() {
+			log.Fatal("client disconnected")
+		}
+
+		freq := float32(machine.CPUFrequency()) / 1000000
+		payload := fmt.Sprintf("%.02fMhz", freq)
+
+		pubVar.PacketIdentifier++
+		err = client.PublishPayload(pubFlags, pubVar, []byte(payload))
+		if err != nil {
+			log.Fatal("error transmitting message: ", err)
+		}
+
+		time.Sleep(time.Second)
+
+		err = client.HandleNext()
+		if err != nil {
+			log.Fatal("handle next: ", err)
```

Comment:
Maybe this is not an ideal use of `HandleNext` as this could block during a `conn.Read()` call until a packet is received. This could potentially deadlock if the `net.Conn` has no timeout set.

Ideally `HandleNext()` is called in its own goroutine. I understand this is limiting for certain microcontrollers like AVR architecture... but those uCs would not have a TCP stack anyways

Reply:
We can add a read deadline on the transport:

```
+               conn.SetReadDeadline(time.Now().Add(10*time.Second))
                err = client.HandleNext()
                if err != nil {
                        log.Fatal("handle next: ", err)
```

I'm not sure what timeout value to use.  1 second is good enough for my testing, but I see 10secs used for timeout in earlier contexts, so using 10sec.

For testing, I tried setting the deadline to 1 Millisecond.  I get the expected timeout i/o failure:

```
ClientId: tinygo-client-GMWJGSAPGA
Connecting to MQTT broker at test.mosquitto.org:1883
Subscribed to topic cpu/freq
1970/01/01 00:00:14 handle next: read tcp <nil>->91.121.93.94:1883: i/o timeout
```

Does this solution look good to you?

Reply:
Hmmm- as a simple example I think it's fine as is. A better alternative would be to leave a link to more complete examples at the repository for natiu: https://github.com/soypat/natiu-mqtt